### ISSUE TITLE
Sparse Matrix CSR format for AtomList/CellList

### ISSUE THREAD

**Issue by ashermancinelli** [Electrostatics/apbs#51]
**Is your feature request related to a problem? Please describe.**
The AtomList and CellList classes are essentially methods on top of a sparsely populated matrix, and the data access patterns for these classes could be vectorized by laying them out linearly in memory.

**Describe the solution you'd like**
We should store all data in AtomList and CellList in COO sparse matrix format, e.g.
```python
class AtomList:
    def __init__(self, natoms: int):
        self.xs = np.array(natoms)
        self.ys = np.array(natoms)
        self.zs = np.array(natoms)
        self.vals = np.array(natoms)
        self.natoms = natoms

    def append(self, a: Atom) -> None:
        self.xs[self.natoms] = a.x
        self.ys[self.natoms] = a.y
        self.zs[self.natoms] = a.z
        self.vals[self.natoms] = a.val
        self.natoms += 1
```
In this case, we would be able to vectorize all operations that work on every value of the atom/cell list, and we could even take advantage of tensor libraries like pytorch or tensorflow and run some code on a GPU.

From what I've seen in the original APBS source, data is either laid out linearly and the index is calculated with an offset, or the data is stored as an array of fields which have the x/y/z coordinates and values as member variables. A sparse matrix format would provide us with similar algorithms at worst, and more vectorization at best. If the atoms usually densely populate the given space, I think we should then use a dense matrix instead.

What are your thoughts on this @sobolevnrm ?

CC @intendo

**Comment by sobolevnrm** [comment]
Since the (x, y, z) coordinates are implictly ordered, I think it might be better to create a N by 3 array with something like:

```
self.coords = np.empty((natoms, 3))
self.coords[:] = np.NaN
```

**Comment by ashermancinelli** [comment]
The data structures seem pretty entwined with the original code - better leave this be for now I think.
